packages/react-app/src/components/giveaways/ParticipationButton.tsx

```diff
@@ -0,0 +1,144 @@
+  const [participationState, setParticipationState] =
+    useState<ParticipationState>(ParticipationState.CHECKING);
+
+  const userInfo = useUserInfo();
+  const [error, setError] = useState(false);
+  const [showPendingModal, setShowPendingModal] = useState(false);
+  const [showRejectedModal, setShowRejectedModal] = useState(false);
+  const [showWinnerModal, setShowWinnerModal] = useState(false);
+
+  const notifyRejection = useCallback(async () => {
+    if (!userInfo) return;
+    await FrontendApiClient.setNotifiedParticipant(
+      giveaway._id,
+      userInfo.email
+    );
+    setShowRejectedModal(true);
+  }, [giveaway, userInfo]);
+
+  const notifyWinner = useCallback(async () => {
+    if (!userInfo) return;
+    await FrontendApiClient.setNotifiedParticipant(
+      giveaway._id,
+      userInfo.email
+    );
+    setShowWinnerModal(true);
+  }, [giveaway, userInfo]);
+
+  const updateParticipationState = useCallback(
+    (fromEvent = false) => {
+      ParticipationService.getParticipationState(
+        giveaway,
+        undefined,
+        userInfo
+      ).then((state) => {
+        setParticipationState(state);
+        onStateChange?.(state);
+
+        if (fromEvent) {
+          if (state === ParticipationState.PENDING) {
+            setShowPendingModal(true);
+          }
+        }
+
+        if (state === ParticipationState.REJECTED) {
+          notifyRejection();
+        }
+
+        ParticipationService.shouldNotifyWinner(giveaway, userInfo).then(
+          (notify) => {
+            if (notify) {
+              notifyWinner();
+            }
+          }
+        );
+      });
+    },
+    [giveaway, userInfo, notifyRejection, notifyWinner, onStateChange]
+  );
+
+  const ActionButton: React.ReactNode = useMemo(() => {
+    let props = {};
+
+    if (participationState === ParticipationState.ALLOWED) {
+      props = {
+        giveaway: giveaway,
+        userInfo: userInfo,
+        successCallback: () => {
+          updateParticipationState(true);
+        },
+        errorCallback: () => {
+          setError(true);
+          updateParticipationState(true);
+        },
+      };
+    } else if (participationState === ParticipationState.MANAGE) {
+      props = { giveaway: giveaway };
+    }
+
+    return React.createElement(
+      ActionButtonComponents[participationState],
+      props
+    );
+  }, [participationState, giveaway, userInfo, updateParticipationState]);
+
+  useEffect(() => {
+    updateParticipationState();
+  }, [updateParticipationState]);
+
+  const closeError = () => {
+    setError(false);
+  };
+
+  return (
+    <div>
+      <Snackbar open={error} autoHideDuration={6000} onClose={closeError}>
+        <MuiAlert severity="error" onClose={closeError}>
+          Oops, something went wrong! Please try again later.
+        </MuiAlert>
+      </Snackbar>
+      <PendingLocationModal
+        open={showPendingModal}
+        onClose={() => setShowPendingModal(false)}
+      />
```

Comment:
Since the visibility is managed by the current component, the modal doesn't need to know if it is open or not:
```suggestion
      {showPendingModal &&
        <PendingLocationModal onClose={() => setShowPendingModal(false)}  />}
```